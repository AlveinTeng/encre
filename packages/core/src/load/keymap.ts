import camelcase from 'camelcase';
import decamelize from 'decamelize';

import type { Opaque } from 'type-fest';

export type RecordId = Opaque<string, 'RecordId'>;

export enum RecordType {
  EVENT = 'event',
  TEMPLATE = 'template',
}

export type SerializedFields = {
  [key: string]: unknown;
};

export type EventSerializedFields = SerializedFields & {
  _recordType: RecordType;
  _secrets: SerializedFields;
  _inputs: SerializedFields;
  _outputs: SerializedFields;
  _parent?: RecordId;
};

export type TemplateSerializedFields = SerializedFields & {
  _recordType: RecordType;
  _children: RecordId[];
};

export type RecordSerializedFields =
  | EventSerializedFields
  | TemplateSerializedFields;

export type SecretFields = {
  [key: string]: string;
};

export type SerializedKeyAlias = {
  [key: string]: string;
};

export function keyToJson(key: string, map?: SerializedKeyAlias): string {
  return map?.[key] || decamelize(key);
}

export function keyFromJson(key: string, map?: SerializedKeyAlias): string {
  return map?.[key] || camelcase(key);
}

export function mapKeys(
  fields: SerializedFields,
  mapper: typeof keyFromJson,
  map?: SerializedKeyAlias
): SerializedFields {
  const mapToKeys: SerializedFields = {};

  for (const key in fields) {
    if (Object.hasOwn(fields, key)) {
      mapToKeys[mapper(key, map)] = fields[key];
    }
  }

  return mapToKeys;
}

export function mapKeyTypes(
  fields: SerializedFields,
  mapper: typeof keyFromJson,
  map?: SerializedKeyAlias
): SerializedFields {
  const mapToKeyTypes: SerializedFields = {};

  for (const key in fields) {
    if (Object.hasOwn(fields, key)) {
      mapToKeyTypes[mapper(key, map)] =
        typeof fields[key] === 'object'
          ? mapKeyTypes(fields[key] as SerializedFields, mapper, map)
          : typeof fields[key];
    }
  }

  return mapToKeyTypes;
}
