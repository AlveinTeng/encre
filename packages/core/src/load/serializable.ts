import { IdProvider } from '../utils/nanoid.js';
// import { nanoid } from 'nanoid';
import {
  type SerializedKeyAlias,
  type SecretFields,
  type SerializedFields,
  mapKeys,
  keyToJson,
  type EventSerializedFields,
  type TemplateSerializedFields,
  mapKeyTypes,
  RecordType,
  type RecordId,
} from './keymap.js';

export enum SerializableType {
  RECORD = 'record',
  MODULE = 'module',
}

export type BasedSerialized<T extends string> = {
  _grp: number;
  _type: T;
  _id: string[];
};

export interface SerializedConstructor extends BasedSerialized<'constructor'> {
  _kwargs: SerializedFields;
}

export interface SerializedSecret extends BasedSerialized<'secret'> {}

export interface SerializedNotImplemented
  extends BasedSerialized<'not_implemented'> {}

export interface SerializedErrnoRecord extends BasedSerialized<'errno_record'> {
  _recordId: RecordId;
}

export interface SerializedInputRecord extends BasedSerialized<'input_record'> {
  _recordId: RecordId;
  _kwargs: SerializedFields;
}

export interface SerializedOutputRecord
  extends BasedSerialized<'output_record'> {
  _recordId: RecordId;
  _kwargs: SerializedFields;
}

export interface SerializedSecretRecord
  extends BasedSerialized<'secret_record'> {
  _recordId: RecordId;
}

export interface SerializedEventRecord extends BasedSerialized<'event_record'> {
  _recordId: RecordId;
  _kwargs: SerializedFields;
  _metadata: EventSerializedFields;
}

export interface SerializedTemplateRecord
  extends BasedSerialized<'template_record'> {
  _recordId: RecordId;
  _metadata: TemplateSerializedFields;
}

export type SerializedRecord =
  | SerializedErrnoRecord
  | SerializedInputRecord
  | SerializedOutputRecord
  | SerializedSecretRecord
  | SerializedEventRecord
  | SerializedTemplateRecord;

export type Serialized =
  | SerializedConstructor
  | SerializedSecret
  | SerializedNotImplemented
  | SerializedRecord;

function shallowCopy<T extends object>(obj: T): T {
  return Array.isArray(obj) ? ([...obj] as T) : ({ ...obj } as T);
}

export function safeAssign<T extends object>(target: T, source: T): T {
  for (const key in source) {
    if (!Object.prototype.hasOwnProperty.call(target, key)) {
      target[key] = source[key];
    }
  }

  return target;
}

export abstract class Serializable {
  _isSerializable = false;

  _kwargs: SerializedFields;

  abstract _namespace: string[];

  static _name(): string {
    return this.name;
  }

  get _id(): string[] {
    return [
      ...this._namespace,
      getUniqueName(this.constructor as typeof Serializable),
    ];
  }

  get _secrets(): SecretFields | undefined {
    return undefined;
  }

  get _attributes(): SerializedFields | undefined {
    return undefined;
  }

  get _aliases(): SerializedFields | undefined {
    return undefined;
  }

  private _idProvider: IdProvider;

  constructor(kwargs?: SerializedFields, ..._args: never[]) {
    this._kwargs = kwargs || {};
    this._idProvider = new IdProvider();
  }

  private _initKwargs(): SerializedFields {
    return Object.keys(this._kwargs).reduce((accumulator, key) => {
      accumulator[key] =
        key in this ? this[key as keyof this] : this._kwargs[key];

      return accumulator;
    }, {} as SerializedFields);
  }

  private async _getRecordId(): Promise<RecordId> {
    const nanoId = await this._idProvider.provideNanoId();

    return nanoId as RecordId;
  }

  private _replaceSecret(
    root: SerializedFields,
    secretsMap: SecretFields
  ): SerializedFields {
    const result: SerializedFields = shallowCopy(root);

    for (const [path, secretKey] of Object.entries(secretsMap)) {
      const [last, ...partsReverse] = path.split('.').reverse();

      let current: SerializedFields = result;
      for (const key of partsReverse.reverse()) {
        if (current[key] === undefined) {
          break;
        }

        current[key] = shallowCopy(current[key] as SerializedFields);
        current = current[key] as SerializedFields;
      }

      if (current[last] !== undefined) {
        current[last] = this.toJSONSecret(secretKey);
      }
    }

    return result;
  }

  private _removeSecret(
    root: SerializedFields,
    secretsMap: SecretFields
  ): SerializedFields {
    const result: SerializedFields = shallowCopy(root);

    for (const [path, secretKey] of Object.entries(secretsMap)) {
      const [last, ...partsReverse] = path.split('.').reverse();

      let current: SerializedFields = result;
      for (const key of partsReverse.reverse()) {
        if (current[key] === undefined) {
          break;
        }

        current[key] = shallowCopy(current[key] as SerializedFields);
        current = current[key] as SerializedFields;
      }

      if (current[last] !== undefined) {
        if (path.split('.').length > 1) {
          delete result[path.split('.')[0]];
        } else {
          delete current[last];
        }
      }
    }

    return result;
  }

  private async _getSecretRecord(
    root: SerializedFields,
    secretsMap: SecretFields
  ): Promise<SerializedFields> {
    const result: SerializedFields = shallowCopy(root);
    const secretKeys: string[] = [];

    for (const [path, secretKey] of Object.entries(secretsMap)) {
      const [last, ...partsReverse] = path.split('.').reverse();

      let current: SerializedFields = result;
      for (const key of partsReverse.reverse()) {
        if (current[key] === undefined) {
          break;
        }

        current[key] = shallowCopy(current[key] as SerializedFields);
        current = current[key] as SerializedFields;
      }

      if (current[last] !== undefined) {
        secretKeys.push(secretKey);
      }
    }

    return (
      secretKeys.length > 0 ? await this.toSecretRecord(secretKeys) : {}
    ) as SerializedFields;
  }

  toJSONNotImplemented(): Serialized {
    return {
      _grp: 1,
      _type: 'not_implemented',
      _id: this._id,
    };
  }

  toJSONSecret(secretKey: string): Serialized {
    return {
      _grp: 1,
      _type: 'secret',
      _id: [secretKey],
    };
  }

  toJSONConstructor(
    aliases: SerializedKeyAlias,
    secrets: SecretFields,
    kwargs: SerializedFields
  ): Serialized {
    return {
      _grp: 1,
      _type: 'constructor',
      _id: this._id,
      _kwargs: mapKeys(
        Object.keys(secrets).length
          ? this._replaceSecret(kwargs, secrets)
          : kwargs,
        keyToJson,
        aliases
      ),
    };
  }

  async toJSONErrnoRecord(): Promise<Serialized> {
    return {
      _grp: 2,
      _type: 'errno_record',
      _id: this._id,
      _recordId: await this._getRecordId(),
    };
  }

  async toSecretRecord(secretKey: string | string[]): Promise<Serialized> {
    return {
      _grp: 2,
      _type: 'secret_record',
      _id: [...secretKey],
      _recordId: await this._getRecordId(),
    };
  }

  async toInputRecord(
    aliases: SerializedKeyAlias,
    secrets: SecretFields,
    kwargs: SerializedFields
  ): Promise<Serialized> {
    return {
      _grp: 2,
      _type: 'input_record',
      _id: this._id,
      _recordId: await this._getRecordId(),
      _kwargs: mapKeyTypes(
        Object.keys(secrets).length
          ? this._removeSecret(kwargs, secrets)
          : kwargs,
        keyToJson,
        aliases
      ),
    };
  }

  async toOutputRecord(
    aliases: SerializedKeyAlias,
    secrets: SecretFields,
    kwargs: SerializedFields
  ): Promise<Serialized> {
    return {
      _grp: 2,
      _type: 'output_record',
      _id: this._id,
      _recordId: await this._getRecordId(),
      _kwargs: mapKeyTypes(
        Object.keys(secrets).length
          ? this._removeSecret(kwargs, secrets)
          : kwargs,
        keyToJson,
        aliases
      ),
    };
  }

  async toEventRecord(
    aliases: SerializedKeyAlias,
    secrets: SecretFields,
    kwargs: SerializedFields,
    outputs: SerializedFields,
    parent?: RecordId
  ): Promise<Serialized> {
    return {
      _grp: 2,
      _type: 'event_record',
      _id: this._id,
      _recordId: await this._getRecordId(),
      _kwargs: mapKeys(
        Object.keys(secrets).length
          ? this._replaceSecret(kwargs, secrets)
          : kwargs,
        keyToJson,
        aliases
      ),
      _metadata: {
        _recordType: RecordType.EVENT,
        _secrets: await this._getSecretRecord(kwargs, secrets),
        _inputs: (await this.toInputRecord(
          aliases,
          secrets,
          kwargs
        )) as SerializedFields,
        _outputs: (await this.toOutputRecord(
          aliases,
          secrets,
          outputs
        )) as SerializedFields,
        _parent: parent,
      },
    };
  }

  async toTemplateRecord(children?: RecordId[]): Promise<Serialized> {
    return {
      _grp: 2,
      _type: 'template_record',
      _id: this._id,
      _recordId: await this._getRecordId(),
      _metadata: {
        _recordType: RecordType.TEMPLATE,
        _children: children ?? [],
      },
    };
  }

  async toRecord(
    outputs: SerializedFields,
    parent?: RecordId
  ): Promise<Serialized> {
    if (!this._isSerializable) {
      return this.toJSONErrnoRecord();
    }

    if (
      this._kwargs instanceof Serializable ||
      typeof this._kwargs !== 'object' ||
      Array.isArray(this._kwargs)
    ) {
      return this.toJSONErrnoRecord();
    }

    let aliases: SerializedKeyAlias = {};
    let secrets: SecretFields = {};
    let kwargs: SerializedFields = this._initKwargs();

    for (
      let sub = Object.getPrototypeOf(this);
      sub;
      sub = Object.getPrototypeOf(sub)
    ) {
      aliases = safeAssign<SerializedKeyAlias>(
        aliases,
        Reflect.get(sub, '_aliases', this)
      );
      secrets = safeAssign<SecretFields>(
        secrets,
        Reflect.get(sub, '_secrets', this)
      );
      kwargs = safeAssign<SerializedFields>(
        kwargs,
        Reflect.get(sub, '_attributes', this)
      );
    }

    Object.keys(secrets).forEach((keyPath: string) => {
      // eslint-disable-next-line @typescript-eslint/no-this-alias, @typescript-eslint/no-explicit-any
      let read: Serializable = this;
      let write: SerializedFields = kwargs;

      const [last, ...partsReverse] = keyPath.split('.').reverse();
      for (const key of partsReverse.reverse()) {
        if (!(key in read) || read[key] === undefined) return;

        if (!(key in write) || write[key] === undefined) {
          if (typeof read[key] === 'object' && read[key] != null) {
            write[key] = {};
          } else if (Array.isArray(read[key])) {
            write[key] = [];
          }
        }

        read = read[key];
        write = write[key] as SerializedFields;
      }

      if (last in read && read[last] !== undefined) {
        write[last] = write[last] || read[last];
      }
    });

    return this.toEventRecord(aliases, secrets, kwargs, outputs, parent);
  }

  toJSON(): Serialized {
    if (!this._isSerializable) {
      return this.toJSONNotImplemented();
    }

    if (
      this._kwargs instanceof Serializable ||
      typeof this._kwargs !== 'object' ||
      Array.isArray(this._kwargs)
    ) {
      return this.toJSONNotImplemented();
    }

    let aliases: SerializedKeyAlias = {};
    let secrets: SecretFields = {};
    let kwargs: SerializedFields = this._initKwargs();

    for (
      let sub = Object.getPrototypeOf(this);
      sub;
      sub = Object.getPrototypeOf(sub)
    ) {
      aliases = safeAssign<SerializedKeyAlias>(
        aliases,
        Reflect.get(sub, '_aliases', this)
      );
      secrets = safeAssign<SecretFields>(
        secrets,
        Reflect.get(sub, '_secrets', this)
      );
      kwargs = safeAssign<SerializedFields>(
        kwargs,
        Reflect.get(sub, '_attributes', this)
      );
    }

    Object.keys(secrets).forEach((keyPath: string) => {
      // eslint-disable-next-line @typescript-eslint/no-this-alias, @typescript-eslint/no-explicit-any
      let read: Serializable = this;
      let write: SerializedFields = kwargs;

      const [last, ...partsReverse] = keyPath.split('.').reverse();
      for (const key of partsReverse.reverse()) {
        if (!(key in read) || read[key] === undefined) return;

        if (!(key in write) || write[key] === undefined) {
          if (typeof read[key] === 'object' && read[key] != null) {
            write[key] = {};
          } else if (Array.isArray(read[key])) {
            write[key] = [];
          }
        }

        read = read[key];
        write = write[key] as SerializedFields;
      }

      if (last in read && read[last] !== undefined) {
        write[last] = write[last] || read[last];
      }
    });

    return this.toJSONConstructor(aliases, secrets, kwargs);
  }
}

export function getUniqueName(serializableClass: typeof Serializable): string {
  const parentClass = Object.getPrototypeOf(serializableClass);
  const isSubclassed =
    typeof serializableClass._name === 'function' &&
    (typeof parentClass._name !== 'function' ||
      serializableClass._name !== parentClass._name);

  if (isSubclassed) return serializableClass._name();

  return serializableClass.name;
}
