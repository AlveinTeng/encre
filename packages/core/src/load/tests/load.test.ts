import { test, expect } from '@jest/globals';
import { stringify } from 'yaml';

import { OptionalImportMap, SecretMap } from '../importType.js';
import { load } from '../index.js';
import { SecretFields, SerializedFields } from '../keymap.js';
import { Serializable } from '../serializable';

test('test custom serializable', async () => {
  class Test extends Serializable {
    _namespace: string[] = ['tests'];

    _isSerializable = true;

    get _secrets(): SecretFields | undefined {
      return { apiKey: 'TEST_API_KEY' };
    }

    get _attributes(): SerializedFields | undefined {
      return { hello: this.hello };
    }

    hello = 3;

    constructor(fields: { aField: string; apiKey: string; hello?: number }) {
      super(fields);
    }
  }

  class SubTest extends Test {
    get _secrets(): SecretFields | undefined {
      return {
        newApiKey: 'NEW_API_KEY',
        apiKey: 'CANNOT_OVERRIDE_API_KEY',
        inheritApiKey: 'INHERIT_API_KEY',
        'nested.api.key': 'NESTED_API_KEY',
      };
    }

    get _attributes(): SerializedFields | undefined {
      return { newAttr: this.newAttr };
    }

    newAttr = 42;

    inheritApiKey: string;

    nested: { api: { key: string } };

    constructor(fields: {
      aField: string;
      apiKey: string;
      newApiKey: string;
      inheritApiKey?: string;
      nested?: { api: { key: string } };
      hello?: number;
      newAttr?: number;
    }) {
      super(fields);

      this.inheritApiKey = fields.inheritApiKey ?? 'this-is-default-key';
      this.nested = fields.nested ?? {
        api: { key: 'this-is-default-nested-key' },
      };
    }
  }

  // In the serializable class, we test the following cases:
  // 1. attributes are decamalized (e.g. aField -> a_field, etcs.);
  // 2. attributes matchs kwargs after convert to JSON (e.g. a_fireld etcs.);
  // 3. loading serializable class (e.g. load<Test>(...))
  const test = new Test({ aField: 'hello', apiKey: 'this-is-a-key' });
  const argumentsBefore = test._kwargs;
  const serializedStr: string = JSON.stringify(test, null, 2);

  expect(test._kwargs).toEqual(argumentsBefore);
  expect(stringify(JSON.parse(serializedStr))).toMatchSnapshot();

  const test2 = await load<Test>(
    serializedStr,
    { TEST_API_KEY: 'this-is-a-key' } as SecretMap,
    { tests: { Test } } as OptionalImportMap
  );

  expect(test2).toBeInstanceOf(Test);
  expect(JSON.stringify(test2, null, 2)).toBe(serializedStr);

  // In the class that extends the parent serializable class,
  // we test the following cases:
  // 1. Inherit all attributes and secrets in the parent class (e.g. aField, etcs.);
  // 2. Cannot override the secret keys (e.g. apiKey in parent requries TEST_API_KEY);
  // 3. Add new attributes (e.g. newAttr etcs.);
  // 4. Nested secret (e.g. nested.api.key);
  // 5. default secret key values (e.g. inheritApiKey, etcs.).
  const subTest = new SubTest({
    aField: 'hello',
    apiKey: 'this-is-a-key',
    newApiKey: 'this-is-another-key',
  });

  const serializedStr2 = JSON.stringify(subTest, null, 2);
  expect(stringify(JSON.parse(serializedStr2))).toMatchSnapshot();

  const subTest2 = await load<Test>(
    serializedStr2,
    {
      CANNOT_OVERRIDE_API_KEY: 'this-is-a-key',
      TEST_API_KEY: 'this-is-a-key',
      NEW_API_KEY: 'this-is-another-key',
      INHERIT_API_KEY: 'this-is-default-key',
      NESTED_API_KEY: 'this-is-default-nested-key',
    } as SecretMap,
    { tests: { SubTest } } as OptionalImportMap
  );

  expect(subTest2).toBeInstanceOf(SubTest);
  expect(JSON.stringify(subTest2, null, 2)).toBe(serializedStr2);
});
