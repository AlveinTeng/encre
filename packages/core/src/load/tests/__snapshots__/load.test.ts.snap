// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test custom module serializable 1`] = `
"_grp: 1
_type: constructor
_id:
  - tests
  - Test
_kwargs:
  a_field: hello
  api_key:
    _grp: 1
    _type: secret
    _id:
      - TEST_API_KEY
  hello: 3
"
`;

exports[`test custom module serializable 2`] = `
"_grp: 1
_type: constructor
_id:
  - tests
  - SubTest
_kwargs:
  a_field: hello
  api_key:
    _grp: 1
    _type: secret
    _id:
      - OVERRIDE_API_KEY
  new_api_key:
    _grp: 1
    _type: secret
    _id:
      - NEW_API_KEY
  new_attr: 42
  hello: 3
  inherit_api_key:
    _grp: 1
    _type: secret
    _id:
      - INHERIT_API_KEY
  nested:
    api:
      key:
        _grp: 1
        _type: secret
        _id:
          - NESTED_API_KEY
"
`;

exports[`test custom record serializable 1`] = `
Object {
  "_grp": 2,
  "_id": Array [
    "tests",
    "Test",
  ],
  "_kwargs": Object {
    "a_field": "hello",
    "api_key": Object {
      "_grp": 1,
      "_id": Array [
        "TEST_API_KEY",
      ],
      "_type": "secret",
    },
    "hello": 3,
  },
  "_metadata": Object {
    "_inputs": Object {
      "_grp": 2,
      "_id": Array [
        "tests",
        "Test",
      ],
      "_kwargs": Object {
        "a_field": "string",
        "hello": "number",
      },
      "_recordId": Any<String>,
      "_type": "input_record",
    },
    "_outputs": Object {
      "_grp": 2,
      "_id": Array [
        "tests",
        "Test",
      ],
      "_kwargs": Object {
        "a_output": "string",
      },
      "_recordId": Any<String>,
      "_type": "output_record",
    },
    "_recordType": "event",
    "_secrets": Object {
      "_grp": 2,
      "_id": Array [
        "TEST_API_KEY",
      ],
      "_recordId": Any<String>,
      "_type": "secret_record",
    },
  },
  "_recordId": Any<String>,
  "_type": "event_record",
}
`;

exports[`test custom record serializable 2`] = `
Object {
  "_grp": 2,
  "_id": Array [
    "tests",
    "SubTest",
  ],
  "_kwargs": Object {
    "a_field": "hello",
    "api_key": Object {
      "_grp": 1,
      "_id": Array [
        "OVERRIDE_API_KEY",
      ],
      "_type": "secret",
    },
    "hello": 3,
    "inherit_api_key": Object {
      "_grp": 1,
      "_id": Array [
        "INHERIT_API_KEY",
      ],
      "_type": "secret",
    },
    "nested": Object {
      "api": Object {
        "key": Object {
          "_grp": 1,
          "_id": Array [
            "NESTED_API_KEY",
          ],
          "_type": "secret",
        },
      },
    },
    "new_api_key": Object {
      "_grp": 1,
      "_id": Array [
        "NEW_API_KEY",
      ],
      "_type": "secret",
    },
    "new_attr": 42,
  },
  "_metadata": Object {
    "_inputs": Object {
      "_grp": 2,
      "_id": Array [
        "tests",
        "SubTest",
      ],
      "_kwargs": Object {
        "a_field": "string",
        "hello": "number",
        "new_attr": "number",
      },
      "_recordId": Any<String>,
      "_type": "input_record",
    },
    "_outputs": Object {
      "_grp": 2,
      "_id": Array [
        "tests",
        "SubTest",
      ],
      "_kwargs": Object {
        "a_output": "string",
      },
      "_recordId": Any<String>,
      "_type": "output_record",
    },
    "_recordType": "event",
    "_secrets": Object {
      "_grp": 2,
      "_id": Array [
        "NEW_API_KEY",
        "OVERRIDE_API_KEY",
        "INHERIT_API_KEY",
        "NESTED_API_KEY",
      ],
      "_recordId": Any<String>,
      "_type": "secret_record",
    },
  },
  "_recordId": Any<String>,
  "_type": "event_record",
}
`;
